monai/transforms/compose.py

```diff
@@ -168,3 +168,82 @@ def inverse(self, data):
             data = apply_transform(t.inverse, data, self.map_items, self.unpack_items)
         return data
+
+
+class OneOf(Compose):
+    """
+    ``OneOf`` provides the ability to radomly choose one transform out of a
+    list of callables with predfined probabilities for each.
+
+    Args:
+        transforms: sequence of callables.
+        weights: probabilities corresponding to each callable in transforms.
+            Probabilities are normalized to sum to one.
+
+    OneOf inherits from Compose and uses args map_items and unpack_items in
+    the same way.
+    """
+
+    def __init__(
+        self,
+        transforms: Optional[Union[Sequence[Callable], Callable]] = None,
+        weights: Optional[Union[Sequence[float], float]] = None,
+        map_items: bool = True,
+        unpack_items: bool = False,
+    ) -> None:
+        super().__init__(transforms, map_items, unpack_items)
+        if len(self.transforms) == 0:
+            weights = []
+        elif weights is None or isinstance(weights, float):
+            weights = [1.0 / len(self.transforms)] * len(self.transforms)
+        if len(weights) != len(self.transforms):
+            raise AssertionError("transforms and weights should be same size if both specified as sequences.")
+        self.weights = ensure_tuple(self._normalize_probabilities(weights))
+        self.index = None
+
+    def _normalize_probabilities(self, weights):
+        if len(weights) == 0:
+            return weights
+        else:
+            weights = np.array(weights)
+            if np.any(weights < 0):
+                raise AssertionError("Probabilities must be greater than or equal to zero.")
+            if np.all(weights == 0):
+                raise AssertionError("At least one probability must be greater than zero.")
+            weights = weights / weights.sum()
+            return list(weights)
+
+    def flatten(self):
+        transforms = []
+        weights = []
+        for t, w in zip(self.transforms, self.weights):
+            # if nested, probability is the current weight multiplied by the nested weights,
+            # and so on recursively
+            if isinstance(t, OneOf):
+                tr = t.flatten()
+                for t_, w_ in zip(tr.transforms, tr.weights):
+                    transforms.append(t_)
+                    weights.append(w_ * w)
+            else:
+                transforms.append(t)
+                weights.append(w)
+        return OneOf(transforms, weights, self.map_items, self.unpack_items)
+
+    def __call__(self, input_):
+        if len(self.transforms) == 0:
+            return input_
+        else:
+            index = self.R.multinomial(1, self.weights).argmax()
+            self.index = index
+            _transform = self.transforms[index]
+            input_ = apply_transform(_transform, input_, self.map_items, self.unpack_items)
+            return input_
+
+    def inverse(self, data):
+        index = self.index
```

Comment:
Hi @rijobro ,

About this inverse comment, do you agree to save the index info in the data dict?

Thanks.

Reply:
Hi @Nic-Ma, do you mean reusing an instance of `OneOf` class multiple times?

Reply:
OneOf is a subclass of ThreadUnsafe, looks like it's fine here

Reply:
Hi @wyli @rijobro ,

The issue is that: when we run the `__call__` of OneOf in multi-processing DataLoader, it recorded different random index to `self.index` in **every subprocess**, when we run `inverse()` later, the `self.index` of main process is actually `None`.
It can be easily reproduced by:
```py
def test_dataloader(self):
    trans = OneOf([Inv(), NonInv()])
    dataset = Dataset(data=[1, 2, 3, 4], transform=trans)
    dataloader = DataLoader(dataset=dataset, batch_size=2, num_workers=2)
    for d in dataloader:
        print(d)
        print("self.index =", trans.index)
        inverted_d = [trans.inverse(i) for i in decollate_batch(d)]
        print(inverted_d)
```
The output is:
```
tensor([2.0000, 2.0135])
self.index = None
[2.0, 2.013509511947632]
tensor([4, 5])
self.index = None
[4, 5]
```

Thanks.

Reply:
Hi @wyli , @rijobro ,

What do you think about this issue now? I am not sure whether I described the issue clearly.

Thanks.

Reply:
I've updated the inverse to make it threadsafe - https://github.com/Project-MONAI/MONAI/pull/2551/files#diff-1b439819c9b8b0015db06f96bb75d0ff8525dea5a86f06c40ae76f2a00821af4R232-R244

Reply:
Hi @rijobro ,

Thanks for the quick update, look good to me now!